src/plugins/visualize/public/application/utils/get_top_nav_config.tsx

```diff
@@ -179,6 +181,30 @@ export const getTopNavConfig = (
     stateTransfer.navigateToWithEmbeddablePackage(originatingApp, { state });
   };
 
+  const confirmModal = (
+    <ConfirmModal
+      onConfirm={() => {
+        if (originatingApp) {
+          stateTransfer.navigateToWithEmbeddablePackage(originatingApp);
+        }
+      }}
+      onCancel={() => {}}
+      title={i18n.translate('visualize.confirmModal.title', {
+        defaultMessage: 'Unsaved changes',
+      })}
+      description={i18n.translate('visualize.confirmModal.confirmTextDescription', {
+        defaultMessage: 'Leave Visualize editor with unsaved changes?',
+      })}
+    />
+  );
+
+  const cancelAndReturn = () => {
+    if (!originatingApp) {
+      return;
+    }
+    showConfirmModal(confirmModal, I18nContext);
```

Comment:
In my opinion, showing a confirm modal every time the user cancels doesn't properly cover this use case. This should be a feature added to the Visualize Editor, similar to the way it is handled in Lens:

1. Keep track of the latest state that was saved. You can update this state on save, and on save and return.
2. Use the `onAppLeave` handler from `AppMountParams` to check if the latest saved state matches the current state. If it doesn't, `onAppLeave` comes with a handy `actions.confirm` method so you don't have to build a whole new component for the confirm modal.

This will mean that you will only get the confirmation when you've actually changed something, _and_ that you get a confirmation when simply navigating to another app

Reply:
Good point! I haven't thought of that really. Will look into it.